src/org/labkey/test/tests/core/security/GetReadableContainersAPITest.java

```diff
@@ -0,0 +1,273 @@
+    {
+        final String path = getPathToFolder(projectReadable, foldersReadable);
+        final boolean readable = foldersReadable.length > 0 ? foldersReadable[foldersReadable.length - 1] : projectReadable;
+        _containerHelper.ensureContainer(path);
+
+        if (readable)
+        {
+            _permissions.addMemberToRole(USER, "Reader", MemberType.user, path);
+        }
+        else
+        {
+            _permissions.removeUserRoleAssignment(USER, "Reader", path);
+        }
+
+        containerInfos.add(new ContainerInfo(path, readable, foldersReadable.length + 1));
+    }
+
+    private String getPathToFolder(boolean projectReadable, boolean... foldersReadable)
+    {
+        StringBuilder path = new StringBuilder();
+        path.append(projectReadable ? READABLE_PROJECT : UNREADABLE_PROJECT);
+        for (int i = 0; i < foldersReadable.length; i++)
+        {
+            boolean readable = foldersReadable[i];
+            path.append("/");
+            path.append(readable ? "Readable" : "Unreadable");
+            path.append(i + 1);
+        }
+        return path.toString();
+    }
+
+    private Set<String> getReadableContainers(Boolean includeSubfolders, Integer depth, Object container)
+    {
+        List<String> response = executeGetReadableContainers(includeSubfolders, depth, container, List.class);
+        return response.stream().filter(path -> path.startsWith(PROJECT_PREFIX)).collect(Collectors.toSet());
+    }
+
+    private List<Map<String, String>> getReadableContainersErrors(Boolean includeSubfolders, Integer depth, Object container)
+    {
+        Map response = executeGetReadableContainers(includeSubfolders, depth, container, Map.class);
+        if (response.get("errors") != null && response.get("errors") instanceof List)
+        {
+            return (List<Map<String, String>>) response.get("errors");
+        }
+        else
+        {
+            throw new RuntimeException("Did not find errors in response: \n" + response.toString());
+        }
+    }
+
+    private <T> T executeGetReadableContainers(Boolean includeSubfolders, Integer depth, Object container, Class<T> expectedResultType)
+    {
+        JSONObject config = new JSONObject();
+        if (includeSubfolders != null)
+            config.put("includeSubfolders", includeSubfolders);
+        if (depth != null)
+            config.put("depth", depth);
+        if (container != null)
+            config.put("container", container);
+
+        String script = "var config = arguments[0];\n" +
+                "config.success = callback;\n" +
+                "config.failure = callback;\n" +
+                "LABKEY.Security.getReadableContainers(config);";
+
+        return executeAsyncScript(script, expectedResultType, config);
+    }
+
+    @Override
+    protected BrowserType bestBrowser()
+    {
+        return BrowserType.CHROME;
+    }
+
+    @Override
+    protected String getProjectName()
+    {
+        return null;
```

Comment:
I assume this is never used? What about throwing UnsupportedOperationException?

Reply:
There are places that call it and know to expect a possible `null` (e.g. `BaseWebDriverTest.doCleanup`). I'd be worried about hitting surprise `NPE`s. This has been our standard pattern for tests that don't create a project.